completion summary

jcdyer opened the pull request:
# [OC-3021](https://tasks.opencraft.com/browse/OC-3021)

Implement an API view that returns the average completion for a course.

## Reviewers

- [ ] @bradenmacdonald 

## TODO:

- [x] Merge #839
- [x] Rebase against new edx-solutions master

jcdyer commented:
This has a hidden "feature" (deliberately undocumented) where staff users can choose which student to use for calculating the denominator of the mean calculation with the standard username= query parameter.  It probably won't have any effect since as I understand it, McKinsey doesn't differentiate available content for different students.

bradenmacdonald commented on lms/djangoapps/completion_api/tests/test_views.py:
```diff
@@ -146,3 +149,48 @@ def test_staff_access(self):
+    def setUp(self):
+        super(CourseSummaryViewTestCase, self).setUp()
+        self.learners = [
+            UserFactory.create(username='learner%s' % i, email='learner%s@example.com' % i, password='test_pass')
```
Nit: Can you just use `UserFactor.create()` or is it helpful to specify these values?

bradenmacdonald commented on lms/djangoapps/completion_api/views.py:
```diff
@@ -103,12 +228,122 @@ def get(self, request):
+
+              * course_key (CourseKey): The unique course identifier.
+              * usage_key: (UsageKey) The unique block identifier.
+              * completion: A dictionary comprising the following fields.
```
Optional: You could just say "see above" here, since the format is the same as `completion` described under "Standard Fields".

bradenmacdonald commented on lms/djangoapps/completion_api/views.py:
```diff
@@ -103,12 +228,122 @@ def get(self, request):
+    **Parameters**
+
+        username (optional):
+            The username of the specified user for whome the course data
```
Typo: `whome`

bradenmacdonald replied:
^ @jcdyer this typo is still in the diff, here and on line 139 too.

bradenmacdonald commented:
@jcdyer Thanks! This looks nice and the new documentation is excellent. One request though, and I'm sorry for not making this more explicit earlier: Could we make the `mean` data so that it is just another value for `requested_fields` in the existing `/api/completion/v0/course/` endpoint?

The mobile team would like to reduce the number of API calls they need to make, so combining these into the same endpoint makes sense.

jcdyer commented:
@bradenmacdonald: I added the summary data to the primary API view.  It required a little reworking to conditionally include it.  I also updated the mean calculation to account for users who don't have a StudentProgress record.  I *think* that will make things more in line with what apros is doing, but haven't had a chance to test it.  I'll need to give this another pass in the morning to compare with Apros, update the view docstrings, and just generally proofread, but I think it's on a good track.

bradenmacdonald commented:
@jcdyer Sounds good, thanks for the update! I'll review it tomorrow then.

jcdyer commented:
@bradenmacdonald This is ready for a review now.

jcdyer commented:
Confirmed that the mean now aligns with the value shown on Apros.  At least for my case.  The denominator is now based on the number of enrollments, not the number of StudentProgress entries (which doesn't include unengaged learners).

bradenmacdonald commented:
@jcdyer When I try now, I'm getting `"mean": 0.05808080808080808` via this API, but Apros is showing '5%' (when I log it out, Apros is using `course_average_value: 5.22727272727`)

To keep things simple, and since it's not worth figuring out which is "more correct" for this interim API, do you think you could just replace your `mean` calculation with a call to [the existing API](https://github.com/edx-solutions/api-integration/blob/680a14d621129165725c262fabc3d09c3d5a911e/edx_solutions_api_integration/courses/views.py#L1860-L1919)? You'd just have to create a fake Request object then call that view directly, and you'll get a result that includes the `course_avg` value.

jcdyer commented:
@bradenmacdonald: I'll need to figure out *how* that view is being called.  It has a lot of options for exclusions and filters that would affect the average calculations (exclude_roles, organizations, groups, etc.), all triggered by query parameters, so unless I call it the same way, we'll still have discrepancies.  I'll spend a little time digging into my test course.

bradenmacdonald commented:
@jcdyer I believe it's normally called with all default parameters. Let me know if you're finding it complex to track down - it should be pretty straightforward to trace starting with the "Progress" page/tab in Apros.

jcdyer commented:
I've got them swapped out.  Tests are currently failing because when I add more users, they aren't getting included in the denominator.    I'm trying to figure out why they are being filtered out.  Maybe I need to add them to an Organization or something.  If you know off the top of your head, I'd love to hear.  Otherwise, I'll be poking around to figure it out.

jcdyer commented:
To be clear, tests are failing, but I'm seeing the right numbers when I access the endpoint and compare against apros.

bradenmacdonald commented:
> Maybe I need to add them to an Organization or something. If you know off the top of your head, I'd love to hear.

No idea off the top of my head, sorry. 

I've triggered a fresh CI build for you.

jcdyer commented:
Sorry.  My tired eyes were wrong.  The number of users did affect it.  The denominator is just different due to the legacy mean calculation using the number of completed assessments, not the number of blocks.

jcdyer commented:
This is ready for another CI run, and hopefully a merge.

bradenmacdonald commented:
CI triggered. We will also need to squash pre-merge.

bradenmacdonald commented:
@jcdyer Simple quality failure. I took the liberty of pushing a fix.

```
lms/djangoapps/completion_api/serializers.py (92.9%):
    13: W0611: (unused-import), : Unused StudentProgress imported from progress.models
```

jcdyer commented:
All's working.  I'm going to squash and rebase.

bradenmacdonald commented on lms/djangoapps/completion_api/views.py:
```diff
@@ -84,10 +86,140 @@ def get_serializer(self):
+
+    **Parameters**
+
+        username (optional):
```
@jcdyer I recently noticed an issue with the interim API: The docstrings, design spec, and course blocks API all use the parameter `username` for this, but the actual code is currently reading the `user` parameter, not `username`. It's not too late to change either the docs or the API itself - should we change it to `username` consistently?

jcdyer replied:
Gah.  Sorry I missed this comment.  

In the database, it's "user" when it is an actual link to the user model, and username when it's just a string.  I think it makes sense to keep it as username in the API, even if that is a mismatch with what's in the model, since the API deals with strings.
